### ISSUE TITLE
Add Ultralytics YOLO integration for computer-vision training workloads

### ISSUE THREAD

**Issue by Pendu** [traceopt-ai/traceml#144]
TraceML currently supports PyTorch Lightning and Hugging Face Trainer for supervised learning. It has no integration for the Ultralytics YOLO training stack — one of the most widely used PyTorch frameworks in computer vision (detection, segmentation, pose, classification). Ultralytics exposes a public callback registry, `model.add_callback(event, func)`, so TraceML can instrument a YOLO run without modifying Ultralytics.

Enable with two lines:

```python
from ultralytics import YOLO
from traceml.integrations.ultralytics import attach_traceml

model = YOLO("yolo11n.pt")
attach_traceml(model)
model.train(data="coco8.yaml", epochs=3)
```

**Scope**

- Add `src/traceml/integrations/ultralytics.py` with `attach_traceml(model)` that registers callbacks via `model.add_callback("on_train_batch_start", …)` and `model.add_callback("on_train_batch_end", …)`.
- Drive TraceML's `trace_step(trainer.model)` across the bracket: enter it in `on_train_batch_start`, exit it in `on_train_batch_end` (the callback receives the `trainer`; `trainer.model` is a `torch.nn.Module`). Import from `traceml.sdk.decorators_compat` (as the HF integration does) so the forward/backward auto-timers are installed.
- Lazily call `trace_model_instance(trainer.model)` once on the first batch for per-layer hooks.
- Respect `TRACEML_DISABLED`; wrap every TraceML call in try/except (see Design notes).
- Add `[ultralytics]` optional extra to `pyproject.toml`.
- Add example `examples/yolo_minimal.py` (`yolo11n` + `coco8`, a few epochs, CPU-runnable).
- Add tests `tests/test_ultralytics_integration.py`.
- Add documentation `docs/user_guide/integrations/ultralytics.md` + an `mkdocs.yml` nav entry under Integrations (next to Hugging Face / PyTorch Lightning).

**Design notes**

- **Callback seam & model reach.** Ultralytics fires `on_train_batch_start` and `on_train_batch_end` once per batch, passing the trainer to each callback (`run_callbacks` → `callback(self)`). Use `trainer.model` — the trained `nn.Module` (it may be `torch.compile`d or DDP-wrapped) — as the target for `trace_step`/`trace_model_instance`.
- **You get per-phase timing, not just step timing.** The forward (`self.model(batch)`), backward (`loss.backward()`), and optimizer step all run *between* `on_train_batch_start` and `on_train_batch_end`. TraceML's `trace_step` installs global forward/backward auto-timers for its span, so wrapping the batch captures forward + backward (+ optimizer on stepping batches) automatically. Mirror the HF integration's `trace_step` usage, but drive `__enter__`/`__exit__` manually across the two callbacks the way the Lightning integration drives its phase contexts.
- **Ultralytics callbacks are fail-CLOSED.** `run_callbacks` does not wrap callbacks in try/except — an exception in a callback propagates and stops training. Wrap every TraceML call in try/except, log to `sys.stderr` with a `[TraceML]` prefix (as the Lightning integration does), and never re-raise into the training loop.
- **Gradient accumulation.** YOLO steps the optimizer only every `accumulate` batches (`if ni - last_opt_step >= self.accumulate`), so not every batch has an optimizer step. Per-batch step timing stays correct; optimizer timing only appears on stepping batches. Consider emitting a 0-duration dummy optimizer event on non-stepping batches to preserve dashboard step alignment, as the Lightning integration does.
- **Deep profile gate.** `trace_model_instance()` is a no-op unless `TRACEML_PROFILE=deep`; per-layer breakdown only appears in deep mode. Document this like the other integrations.
- **CUDA.** The auto-timers handle CUDA timing internally; if you add any manual wall-clock timing, call `torch.cuda.synchronize()` first.
- **Lazy attach once** on the first `on_train_batch_start` (the model id may change with compile/DDP), matching the HF integration.

**Acceptance Criteria**

- [ ] `attach_traceml(model)` then `model.train(...)` runs to completion with TraceML active.
- [ ] Per-batch step timing appears in TraceML's summary; forward and backward phase timings are captured (optimizer timing on stepping batches).
- [ ] Per-layer forward/backward timing + memory appear when `TRACEML_PROFILE=deep`.
- [ ] Works on a CPU-only `coco8` smoke test with `yolo11n` (no large external downloads beyond the tiny weights/dataset Ultralytics auto-fetches).
- [ ] `TRACEML_DISABLED=1` cleanly bypasses all instrumentation.
- [ ] An exception raised inside a TraceML callback is caught + logged, never crashes YOLO training (Ultralytics' `run_callbacks` is fail-closed).
- [ ] Gradient-accumulation runs (`accumulate > 1`) keep correct step alignment.
- [ ] `ultralytics` is an optional extra; importing the integration without it raises a clear `ImportError` (mirror the HF integration's guard).

**References**

- Existing patterns: `src/traceml/integrations/huggingface.py` (coarse, `trace_step` wrap, lazy attach), `src/traceml/integrations/lightning.py` (manual context across callbacks, fail-open logging, gradient-accumulation handling).
- Core SDK: `trace_step`, `trace_model_instance` in `src/traceml/sdk/instrumentation.py` (import via `traceml.sdk.decorators_compat`).
- Ultralytics: `add_callback` (`ultralytics/engine/model.py`), callback firing in `ultralytics/engine/trainer.py`, event list `ultralytics/utils/callbacks/base.py`. Docs: https://docs.ultralytics.com/usage/callbacks

**Comment by Pendu** [comment]
Closing as a duplicate of #132, which was closed earlier for the same reason.

Ultralytics is licensed **AGPL-3.0**, while TraceML is **Apache-2.0**. Shipping an `ultralytics`-importing adapter inside the TraceML package — even behind an optional extra — would make a TraceML release a combined work with AGPL-3.0 code, and AGPL §13's network clause is a particular concern for a hosted/SaaS context. The established ecosystem pattern is that the AGPL-side project hosts the integration glue (Ultralytics ships its own `wb.py` / `mlflow.py` / `comet.py` callbacks under AGPL); permissively-licensed tools do not ship Ultralytics-importing code.

The AGPL-safe ways to support YOLO users, which we'll pursue instead:
1. **Docs-only** — a guide showing users how to wire TraceML's public API into YOLO via `model.add_callback(...)` in their own training script (no TraceML→ultralytics import, no optional extra).
2. **Upstream** — contribute a TraceML callback into `ultralytics/utils/callbacks/`, mirroring the existing logger callbacks.

Thanks to Abhinav for catching the licensing issue.
